src/Lecture4.hs

```diff
@@ -131,8 +137,30 @@ errors. We will simply return an optional result here.
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
 
+trimSpaces :: String -> String
+trimSpaces = reverse . dropWhile isSpace . reverse . dropWhile isSpace
+
+splitOn :: Char -> String -> [String]
+splitOn _ [] = []
+splitOn sym str@(x:xs)
+  | x == sym = splitOn sym xs
+  | otherwise = clearedPart : splitOn sym other
+  where (part, other) = break (==sym) str
+        clearedPart = trimSpaces part
+
+parseCost :: String -> Maybe Int
+parseCost stringCost
+  | isNothing parsedValue = parsedValue
+  | fromJust parsedValue >= 0 = parsedValue
+  | otherwise = Nothing
+  where parsedValue = readMaybe stringCost
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow str
+  | length clearedData == 3 = let (pName:pType:pCost:_) = clearedData in Row pName <$> readMaybe pType <*> parseCost pCost
+  | otherwise = Nothing
+  where clearedData = splitOn ',' str
```

Comment:
I haven't explained this during lectures but there's a nice way to write this parsing syntax using `do`-notation, `Monad` instance for `Maybe` and the `MonadFail` desugarding trick 🧙🏻 

You can read more about this trick here:

* https://kowainik.github.io/posts/haskell-mini-patterns#monadfail-sugar

Applying this programming pattern, the `parseRow` function will look like this:

```suggestion
parseRow str = do
    [pName, pType, pCost] <- Just (splitOn ',' str)
    typ <- readMaybe pType
    cost <- readMaybe pCost
    if cost >= 0
        then Just (Row pName typ cost)
        else Nothing
```